provisioner/data.js

```diff
@@ -86,6 +88,110 @@ var WorkerType = base.Entity.configure({
+    minPrice: base.Entity.types.Number,
+    maxPrice: base.Entity.types.Number,
+    canUseOndemand: base.Entity.types.JSON, // delete this
+    canUseSpot: base.Entity.types.JSON, // delete this
+    instanceTypes: base.Entity.types.JSON,
+    regions: base.Entity.types.JSON,
+    // Store the date of last modification for this entity
+    lastModified: base.Entity.types.Date,
+    // Global base UserData object for overwriting
+    userData: base.Entity.types.JSON,
+    // Global base LaunchSpecification object for overwriting
+    launchSpec: base.Entity.types.JSON,
+    // Global base secrets object for overwriting
+    secrets: base.Entity.types.JSON,
+    // Global base scope list for appending, these are the scopes that
+    // temporary TC credentials will be issued against
+    scopes: base.Entity.types.JSON,
+  },
+  migrate: function (item) {
+    // First, let's set up the static/easy data
+    var newWorker = {
+      workerType: item.workerType,
+      minCapacity: item.minCapacity || 0,
+      maxCapacity: item.maxCapacity,
+      scalingRatio: item.scalingRatio,
+      minPrice: item.minPrice,
+      maxPrice: item.maxPrice,
+      canUseOnDemand: item.canUseOnDemand,
+      canUseSpot: item.canUseSpot,
+      lastModified: new Date(),
+    };
+
+    // We do this three times, lets just stick it into a function
+    function fixUserData (x) {
+      var ud = JSON.parse(new Buffer(x, 'base64').toString());
+      // These keys get overwriten automatically by the provisioner
+      // so we don't really need them to ever be in UserData
+      ['provisionerId', 'workerType', 'capacity'].forEach(y => {
+        delete ud[y];
+      });
+      return ud;
+    }
+
+    // Global User Data
+    if (item.launchSpecification.UserData) {
+      newWorker.userData = fixUserData(item.launchSpecification.UserData);
+    }
+
+    // We want to make sure that the global user data is correct
+    newWorker.launchSpec = lodash.cloneDeep(item.launchSpecification);
+    delete newWorker.launchSpec.UserData;
+
+    // Now we set up defaults for the secrets and scopes
+    newWorker.secrets = {};
+    newWorker.scopes = [];
+
+    // Now let's fix up the regions
+    newWorker.regions = item.regions.map(r => {
+      var region = {
+        region: r.region,
+        secrets: {},
+        scopes: [],
+      };
+
+      region.userData = fixUserData(r.overwrites.UserData);
+      region.launchSpec = lodash.cloneDeep(r.overwrites);
+      delete region.launchSpec.UserData;
+
+      return region;
+    });
+
+    // Now let's fix up the instance types
+    newWorker.instanceTypes = item.instanceTypes.map(t => {
+      var it = {
+        instanceType: t.instanceType,
+        capacity: t.capacity,
+        utility: t.utility,
+        secrets: {},
+        scopes: [],
+      };
+
+      it.userData = fixUserData(t.overwrites.UserData);
+      it.launchSpec = lodash.cloneDeep(t.overwrite);
+      delete it.launchSpec.UserData;
```

Comment:
Why aren't these part of the construct above?

You don't even need to create `it`, you can just do:
```js
return {
  instanceType: t.instanceType,
  capacity: t.capacity,
  utility: t.utility,
  secrets: {},
  scopes: [],
  userData: fixUserData(t.overwrites.UserData),
  launchSpec: lodash.omit(lodash.cloneDeep(t.overwrite), 'UserData')
};
```

Exactly the same goes to the regions transformations above too..